src/libstore/derivations.hh

```diff
@@ -6,17 +6,48 @@
 #include "content-address.hh"
 
 #include <map>
+#include <variant>
 
 
 namespace nix {
 
 
 /* Abstract syntax of derivations. */
 
-struct DerivationOutput
+struct DerivationOutputInputAddressed
 {
+    /* Will need to become `std::optional<StorePath>` once input-addressed
+       derivations are allowed to depend on cont-addressed derivations */
     StorePath path;
-    std::optional<FixedOutputHash> hash; /* hash used for expected hash computation */
+};
+
+struct DerivationOutputFixed
+{
+    FixedOutputHash hash; /* hash used for expected hash computation */
+};
+
+struct DerivationOutputFloating
+{
+    /* information used for expected hash computation */
+    FileIngestionMethod method;
+    HashType hashType;
```

Comment:
Should we actually have this flexibility, or should we just enforce NAR-sha256?

Reply:
If we do remove these fields, we can still encode drv files in a way that allows for adding them later